--- elfutils-0.152-org/libelf/libelfP.h
+++ elfutils-0.152/libelf/libelfP.h
@@ -63,6 +63,19 @@
 #include <stdio.h>
 #include <string.h>
 
+#ifndef MIN
+# define MIN(a,b) (((a)<(b))?(a):(b))
+#endif
+#ifndef MAX
+# define MAX(a,b) (((a)>(b))?(a):(b))
+#endif
+#ifndef powerof2
+# define powerof2(x) ((((x) - 1) & (x)) == 0)
+#endif
+#ifndef __CONCAT
+#define __CONCAT(x,y)   x ## y
+#endif
+
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
