--- a/gcc/cp/cfns.h
+++ b/gcc/cp/cfns.h
@@ -51,10 +51,6 @@
 __inline
 #endif
 static unsigned int hash (const char *, unsigned int);
-#ifdef __GNUC__
-__inline
-#endif
-const char * libc_name_p (const char *, unsigned int);
 /* maximum key range = 391, duplicates = 0 */
 
 #ifdef __GNUC__
@@ -124,11 +120,8 @@
 
 #ifdef __GNUC__
 __inline
-#ifdef __GNUC_STDC_INLINE__
-__attribute__ ((__gnu_inline__))
-#endif
 #endif
-const char *
+static const char *
 libc_name_p (register const char *str, register unsigned int len)
 {
   enum
